packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -740,10 +750,109 @@ export function runConnectedCallback(vm: VM) {
     }
 }
 
+function setupContext(vm: VM) {
+    const contextKeys = getContextKeys();
+
+    if (!contextKeys) {
+        return;
+    }
+
+    const { connectContext, contextEventKey } = contextKeys;
+    const { component } = vm;
+    const enumerableKeys = keys(getPrototypeOf(component));
+    const contextfulFieldsOrProps = enumerableKeys.filter(
+        (propName) => (component as any)[propName]?.[connectContext]
+    );
+
+    if (contextfulFieldsOrProps.length === 0) {
+        return;
+    }
+
+    let isProvidingContext = false;
+    const providedContextVarieties = new Map<unknown, Signal<unknown>>();
+    const contextRuntimeAdapter = {
+        isServerSide: false,
+        component,
+        provideContext<T extends object>(
+            contextVariety: T,
+            providedContextSignal: Signal<unknown>
+        ): void {
+            if (!isProvidingContext) {
+                isProvidingContext = true;
+
+                component.addEventListener(ContextEventName, (event: any) => {
+                    if (
+                        event.detail.key === contextEventKey &&
+                        providedContextVarieties.has(event.detail.contextVariety)
+                    ) {
+                        event.stopImmediatePropagation();
+                        const providedContextSignal = providedContextVarieties.get(
+                            event.detail.contextVariety
+                        );
+                        event.detail.callback(providedContextSignal);
+                    }
+                });
+            }
+
+            let multipleContextWarningShown = false;
+
+            if (providedContextVarieties.has(contextVariety)) {
+                if (!multipleContextWarningShown) {
+                    multipleContextWarningShown = true;
+                    logError(
+                        'Multiple contexts of the same variety were provided. Only the first context will be used.'
+                    );
+                }
```

Comment:
This code is not covered in the Karma tests. You can download the [coverage report](https://github.com/salesforce/lwc/actions/runs/12186089972?pr=4995) and see that actually quite a bit is uncovered.

![Screenshot 2024-12-10 at 9 59 01 AM](https://github.com/user-attachments/assets/7c1547ab-a8ce-4c3c-8692-b075d9633abb)